pkg/ck8s/workload_cluster.go

```diff
@@ -192,58 +206,91 @@ func (w *Workload) GetK8sdProxyForControlPlane(ctx context.Context) (*K8sdClient
+	request := &apiv1.RemoveNodeRequest{Name: machine.Name, Force: true}
+
+	// If we see that ignoring control-planes is causing issues, let's consider removing it.
+	// It *should* not be necessary as a machine should be able to remove itself from the cluster.
+	err := w.doK8sdRequest(ctx, http.MethodPost, "1.0/x/capi/remove-node", request, nil, k8sdProxyOptions{IgnoreNodes: map[string]struct{}{machine.Name: {}}})
 	if err != nil {
-		return "", fmt.Errorf("failed to create k8sd proxy: %w", err)
+		return fmt.Errorf("failed to remove %s from cluster: %w", machine.Name, err)
 	}
+	return nil
+}
+
+func (w *Workload) doK8sdRequest(ctx context.Context, method, endpoint string, request any, response any, k8sdProxyOptions k8sdProxyOptions) error {
 	type wrappedResponse struct {
-		Error    string                     `json:"error"`
-		Metadata apiv1.GetJoinTokenResponse `json:"metadata"`
+		Error    string          `json:"error"`
+		Metadata json.RawMessage `json:"metadata"`
 	}
 
-	requestBody, err := json.Marshal(apiv1.GetJoinTokenRequest{Name: name, Worker: worker})
+	k8sdProxy, err := w.GetK8sdProxyForControlPlane(ctx, k8sdProxyOptions)
+	if err != nil {
+		return fmt.Errorf("failed to create k8sd proxy: %w", err)
+	}
+
+	url := fmt.Sprintf("https://%s:%v/%s", k8sdProxy.NodeIP, w.microclusterPort, endpoint)
+
+	requestBody, err := json.Marshal(request)
 	if err != nil {
-		return "", fmt.Errorf("failed to prepare worker info request: %w", err)
+		return fmt.Errorf("failed to prepare worker info request: %w", err)
 	}
 
-	req, err := http.NewRequestWithContext(ctx, http.MethodPost, fmt.Sprintf("https://%s:%v/1.0/x/capi/generate-join-token", k8sdProxy.NodeIP, microclusterPort), bytes.NewBuffer(requestBody))
+	req, err := http.NewRequestWithContext(ctx, method, url, bytes.NewBuffer(requestBody))
 	if err != nil {
-		return "", fmt.Errorf("failed to create request: %w", err)
+		return fmt.Errorf("failed to create request: %w", err)
 	}
 
-	req.Header.Add("capi-auth-token", authToken)
+	req.Header.Add("capi-auth-token", w.authToken)
 	res, err := k8sdProxy.Client.Do(req)
 	if err != nil {
-		return "", fmt.Errorf("failed to reach k8sd through proxy client: %w", err)
+		return fmt.Errorf("failed to reach k8sd through proxy client: %w", err)
 	}
 	defer res.Body.Close()
 
-	var wrappedResp wrappedResponse
-	if err := json.NewDecoder(res.Body).Decode(&wrappedResp); err != nil {
-		return "", fmt.Errorf("failed to parse HTTP response: %w", err)
+	if response == nil {
+		// Nothing to decode
+		return nil
 	}
-	if res.StatusCode != 200 {
-		return "", fmt.Errorf("request for join token request failed: %s", wrappedResp.Error)
+
+	var responseBody wrappedResponse
+	if err := json.NewDecoder(res.Body).Decode(&responseBody); err != nil {
+		return fmt.Errorf("failed to parse HTTP response: %w", err)
+	}
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("HTTP request failed with status code: %d (%s)", res.StatusCode, responseBody.Error)
+	}
```

Comment:
should we not first check the status coe, then start decoding things (or skip decoding if the response is `nil`)?